asv/environment.py

```diff
@@ -862,11 +863,39 @@ def _build_project(self, repo, commit_hash, build_dir):
         """
         Run build commands
         """
+        build_dir_path = Path(build_dir)
+
+        def has_file(file_name):
+            return (build_dir_path / file_name).exists()
+
         cmd = self._build_command
+
         if cmd is None:
-            cmd = ["python setup.py build",
-                   ("PIP_NO_BUILD_ISOLATION=false "
-                    "python -mpip wheel --no-deps --no-index -w {build_cache_dir} {build_dir}")]
+            if has_file('pyproject.toml'):
+                if has_file('pdm.lock'):
+                    cmd = [
+                        "python -mpip install pdm",
+                        "pdm install --prod",
+                        "pdm build --wheel --no-isolation --output {build_cache_dir}"
+                    ]
+                elif has_file('poetry.lock'):
+                    cmd = [
+                        "python -mpip install poetry",
+                        "poetry install",
+                        "poetry build --format wheel --output {build_cache_dir}"
+                    ]
```

Comment:
The problem is that `poetry` and `pdm` are usually installed system-wide (at the user level with their own install scripts) and not always part of the `requirements.txt`. This is problematic because some projects use `pdm` or `poetry` for dependencies / development / tool management but still have `setuptools` as the build backend, which leads to errors, since `asv` doesn't pick up `pdm` or `poetry` from system-wide installs.

One way around this is to add logic in `asv` to look for `pdm` and `poetry` but this seemed less invasive.